barrier/barrier.1
.\" $Id: barrier.1,v 1.9 2006/01/23 18:17:28 garbled Exp $
.\" Copyright (c) 1998, 1999, 2000
.\"	Tim Rightnour.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgment:
.\"	This product includes software developed by Tim Rightnour.
.\" 4. The name of Tim Rightnour may not be used to endorse or promote 
.\"    products derived from this software without specific prior written 
.\"    permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY TIM RIGHTNOUR ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL TIM RIGHTNOUR BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" The following requests are required for all man pages.
.Dd October 14, 1998
.Dt BARRIER 1
.Sh NAME
.Nm barrier
.Nd syncronize a process on a number of machines.
.Sh SYNOPSIS
.Nm
.Op Fl q
.Op Fl v
.Op Fl h Ar barrier_host
.Op Fl k Ar key
.Op Fl p Ar port
.Fl s Ar cluster_size
.Sh DESCRIPTION
The 
.Nm
command can be used to syncronize execution of various commands.  When
a barrier is set, it is not released until all the nodes or processes
have met the barrier condition.  This can be a handy way to make sure
slower machines, perform certain tasks before doing something on
faster machines that relies on them. The following options are available:
.Bl -tag -width www
.It Fl q
Turns quiet off, so barrier notifications are sent to the user.
.It Fl v
Prints the version of ClusterIt to the stdout, and exits.
.It Fl h
Specifies a host, which is running barrierd, to connect to for
barrier syncronization.  Overrides the
.Ev BARRIER_HOST
environment variable.
.It Fl k
Specifies a unique key to syncronize with.  A barrier will only
synchronize with other barriers that share the same key.  Defaults to
the string 'barrier'.
.It Fl p
Specifies a port number of a remote barrier daemon to connect to.  Defaults
to 1933, and overrides the
.Ev BARRIER_PORT
environment variable.
.It Fl s
Sets the size of the cluster.  A barrier condition is met, when the number of
barrier clients connected to the remote daemon, sharing the same unique key, is
equal to this number.
.El
.Sh ENVIRONMENT
.Nm
utilizes the following environment variables.
.Bl -tag -width "BARRIER_PORT"
.It Ev BARRIER_PORT
Sets the default remote port to connect to.
.It Ev BARRIER_HOST
Sets the default remote host to connect to.
.El
.\" .Sh FILES
.Sh DIAGNOSTICS
Exit status is 0 on successful syncronization.  Exit status will be set to 1 if
the connection is lost, or the server rejects your connection.
.Sh SEE ALSO
.Xr barrierd 1 ,
.Xr dsh 1 ,
.Sh HISTORY
The
.Nm
command appeared in clusterit 1.1. 
.Sh AUTHOR
.Nm Barrier
was written by Tim Rightnour.


CHANGES
$Id: CHANGES,v 1.15 2005/12/13 08:04:25 garbled Exp $

Add a -v option to all programs to tell the user what version of
ClusterIt they are running.  CVS will allways have _BETA tagged on the end,
and it will be removed at release.

Add RVT_CMD_ARGS to dvt, to allow passing of args to rvt

Incorporate patches from debian xvt package into rvt:    
Fix xvt manpage typos
Add -rv (reverse-video) option 
Better colormap handling
All other patches were either irrelevant (makefile fixes, etc) or allready
incorporated.

Fix pdf to work on linux hosts

Replace most instances of sprintf with snprintf. Did not touch rvt, as I would
prefer to keep that as unmodified as possible, and most of them would not
be easy to fix anyhow.

Replace most instances of malloc with calloc.   Mostly to prevent any
unanticipated wierdness, and a few of the mallocs were immediately followed
by memset commands anyhow.

Add RCMD_CMD_ARGS to most of the programs. This makes it easier to do
things like run "ssh -4" as your RCMD_CMD.

Stop allocating the command to be run on the remote end as buf[MAXBUF],
instead compute it at runtime and dynamically allocate it to prevent
possible overflow on a huge command.

Fix a bug noted by Jot Powers, where trailing whitespace on a GROUP or LUMP
entry would cause dsh -g to not match it.

Fix a bug noted by John Bales  where linux machines often recieved truncated 
output from the child ssh/rsh process.

Fix bug where all the programs would mangle argv[0] for ps.

* Release 2.3.1 *

Fix a bug noted by James F Wilkus on OpenBSD where dsh would segv when
reading the clusterfile.

Convert seq over to rseq fully, not just for linux anymore.  Sigh.

Fix the clusterit.spec file to work properly.  Tested on redhat

* Release 2.3 *

Made all remote commands use poll instead of a while loop to read the incoming
stream of data from stdout and stderr.  This fixes bug where linux ssh will
hang up.

Fixed linux command line argument processing.

Changed the -l <username> processing to invoke RCMD_CMD with user@host rather
than -l user host.  This makes it much more reliable.

Added an option to dsh to test remote ports before connecting.  This can be
useful where a remote host might be down, and a long timeout would occur
waiting for a response.

Updated rvt to use a newer set of sources from xvt.  Fixes questionable
licensing in xvt, and makes it more reliable.

Converted makefiles to autoconf for ease of porting.  No more stupid patch
sets which were a terrible burden to maintain and generate.

ANSIfyed most of the code (other than rvt). Nobody uses KNF anymore.. sigh..

Incorporated many patches sent by users from the internet.  I have
unfortunately lost track of what patch came from who, but thier help is
greatly appreciated.  Most of these patches either fixed bugs, added stability
or fixed porting bugs.  The xvt conversion was based mainly on a patch sent
in by a user.  My thanks go out to those of you who have sent them in.

# changes since 2.2

* Release 2.2 *
Missing this data, sorry
* Release 2.0 *

Added dvt and rvt..  an interactive form of dsh with xterms.

Fixed a bug in jsd and dsh where it would lock up running ssh waiting for
input.

Add a jsd dir, and move jsd.c from dsh to the jsd dir.  Add nearly complete
jsh program, and jcommon.[ch].

Add new common subdir and move a tonnage of crap in there.  Mainly
common.c, common.h and strsep.c.

Fix horrid bug in dsh where dsh | more would lock terminals up.  Still
need to fix in seq, run, probably pcp.

Add <netinet/if_ether.h> to barrierd.c to shut AIX C up.

Lots of minor bugfixes, mostly dealing with failures to open files, opening
0 byte clusters, etc etc..

Pcp was still using it's own code to parse strings and groups.  Fully rewrote
the back-end internals to use the new linked-list stuff, and utilize common.c.
Pcp's code is now virtually identical to dsh, whereas before it had diverged
significantly.

Error handling has finally graduated to where I no longer feel that it
"needs work".

Fix a place where malloced data was strcatted to without being bzeroed.

Group, exclusion and -w nodes are now dynamicly generated arrays.  Removes
the MAX_GROUPS restrictions.

The nodelist and grouplists are no longer maintained in any way similar to
the original setup.  Nodes are not stored in a flat array of nodes, with a
corresponding flat array of group names.  Nodes are now stores in a linked
list of machines to run on.  Nodes which are not to be run on, are not
included in this list, so the back-end routines only need to execute code
on every node they are handed in the linked list.  This is all handled in
common.c now.  This removes the all compile time restrictions on MAXNODES,
allowing the program to scale much more efficiently.

# changes since 1.3
* Release 1.3 *

Break dsh, run and seq's common code into a common.c and common.h file,
so I don't forget to add options to run and seq when fixing dsh.  Also, fix
some locations where run and seq were out of sync slightly.

Fix a bug in pdf where if the -m option was used, filesystems at 100%
would not be displayed.

Handle ^C's correctly.  If a remote process is hung, and it is the current
process being read, you may ^C it, and the interrupt will be sent to the
remote process.  This would be just like hitting ^C on the remote end, 
and will only affect the process you are currently reading from.

Added ability to pcp multiple files and wildcards to remote hosts.

Made pcp and dsh compile on an ANSI compiler (IBM C, SUN C)

Make output from dsh look nicer, and format well.

# changes since 1.2
* Release 1.2 *

Bug fixes in dsh to deal with the start of a fangroup, and
a race condition, where large output would cause it to hang
waiting for descriptors to die.

Fix a bug in dsh that caused it to lose data on AIX, and 
dump core when transitioning fangroups.

Added the rungroup, and exclusion commands to all the programs.

Fix pdf to work on a heterogeneous cluster.


barrier/barrier.c
/* $Id: barrier.c,v 1.19 2005/12/13 05:01:54 garbled Exp $ */
/*
 * Copyright (c) 1998, 1999, 2000
 *	Tim Rightnour.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by Tim Rightnour.
 * 4. The name of Tim Rightnour may not be used to endorse or promote 
 *    products derived from this software without specific prior written 
 *    permission.
 *
 * THIS SOFTWARE IS PROVIDED BY TIM RIGHTNOUR ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL TIM RIGHTNOUR BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <stdlib.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netdb.h>
#include <stdio.h>
#include <string.h>

#include "../common/common.h"
#include "../common/sockcommon.h"

#if !defined(lint) && defined(__NetBSD__)
__COPYRIGHT(
"@(#) Copyright (c) 1998, 1999, 2000\n\
        Tim Rightnour.  All rights reserved\n");
__RCSID("$Id: barrier.c,v 1.19 2005/12/13 05:01:54 garbled Exp $");
#endif

int quietflag, barrier_port, debug;
char *barrier_host;
char *progname;

int make_barrier(char *key, int nodes);
int write_to_server(int filedes, char *buf);
void _log_bailout(int line, char *file);

int
main(int argc, char **argv)
{
    extern char *optarg;
    extern char *version;

    char *key, *p, *q;
    int nodes, ch, code;
	
    barrier_port = BARRIER_SOCK;
    barrier_host = NULL;
    quietflag = 1;
    key = NULL;
    nodes = 0;

    progname = p = q = strdup(argv[0]);
    while (progname != NULL) {
	q = progname;
	progname = (char *)strsep(&p, "/");
    }
    progname = q;

#if defined(__linux__)
    while ((ch = getopt(argc, argv, "+?dh:k:p:s:v")) != -1)
#else
    while ((ch = getopt(argc, argv, "?dh:k:p:s:v")) != -1)
#endif
	switch (ch) {
	case 'q':
	    quietflag = 0;
	    break;
	case 'h':
	    barrier_host = strdup(optarg);
	    break;
	case 'k':
	    key = strdup(optarg);
	    break;
	case 'p':
	    barrier_port = atoi(optarg);
	    break;
	case 's':
	    nodes = atoi(optarg);
	    break;
	case '?':
	    (void)fprintf(stderr, "usage: barrier [-qv] [-h barrier_host]"
			  " [-k key] [-p port] -s cluster_size\n");
	    exit(EXIT_FAILURE);
	    break;
	case 'v':
	    (void)printf("%s: %s\n", progname, version);
	    exit(EXIT_SUCCESS);
	    break;
	default:
	    break;
	}
    if (nodes < 1) {
	(void)fprintf(stderr, "usage: barrier [-q] [-h barrier_host] "
	    "[-k key] [-p port] -s cluster_size\n");
	exit(EXIT_FAILURE);
    }
    if (key == NULL)
	key = strdup("barrier");
    if (key == NULL)
	bailout();

    if (barrier_port == BARRIER_SOCK)
	if (getenv("BARRIER_PORT") != NULL)
	    barrier_port = atoi(getenv("BARRIER_PORT"));

    if (barrier_host == NULL) {
	if (getenv("BARRIER_HOST") != NULL)
	    barrier_host = strdup(getenv("BARRIER_HOST"));
	else {
	    (void)fprintf(stderr, "No barrier host given on command line,"
		" and BARRIER_HOST environment not found.\n");
	    exit(EXIT_FAILURE);
	}
    }

    code = make_barrier(key, nodes);
    if (code == 6) {
	if (!quietflag)
	    (void)printf("Barrier met, continuing: %s\n", key);
	return(EXIT_SUCCESS);
    } else
	if (!quietflag)
	    (void)printf("Barrier encountered a server error, aborting.\n");
    return(EXIT_FAILURE); 
}

int make_barrier(char *key, int nodes)
{
    int sock;
    char *p;
    char message[256];
    struct sockaddr_in name;
    struct hostent *hostinfo;

    /* create socket */
    sock = socket(PF_INET, SOCK_STREAM, 0);
    if (sock < 0) {
	perror("socket");
	exit(EXIT_FAILURE);
    }

    name.sin_family = AF_INET;
    name.sin_port = htons(barrier_port);
    hostinfo = gethostbyname(barrier_host);

    if (hostinfo == NULL) {
	(void)fprintf(stderr, "Unknown host %s.\n", barrier_host);
	exit(EXIT_FAILURE);
    }

    name.sin_addr = *(struct in_addr *) hostinfo->h_addr;

    if (connect (sock, (struct sockaddr *) &name, sizeof (name)) < 0) {
	perror ("connect");
	exit (EXIT_FAILURE);
    }
    if (!quietflag)
	(void)printf("Barrier syncing with token: %s\n", key);
    (void)snprintf(message, 256, "%s %d", key, nodes);
    write_to_server(sock, message);
    return(read(sock, &p, 6));
}

int write_to_server(int filedes, char *buf)
{
    int nbytes;

    nbytes = write(filedes, buf, strlen(buf));
    if (nbytes < 0) {
	return(EXIT_FAILURE);
    }
    return(0);
}

void
_log_bailout(int line, char *file)
{
    _bailout(line, file);
}
